chore(agw): integrate dhcp cli helper

crasu opened the pull request:
## Summary

Integrate the cli helper created here  #14418 into mobilityd dhcp allocator

- Added `udhcpd.conf` to create a new traffic server service for testing.
- Added udhcpd service trfserver `main.yml`.
- Added new integration test for non_nat configuration.
- Removed `dhcp_client.py` and moved functionality into `ip_allocator_dhcp.py`.
- Renamed test files to be more consistent/clear about what they are doing.
- Some changes to Bazel so that the new tests are running .
- A number of fixes for `dhcp_helper_cli.py` for issues we uncovered during testing.



## Test Plan

- [x] Added unit tests for `dhcp_helper_cli.py`
- [x] Added new integration test `test_attach_detach_with_non_nat_dhcp.py` and got it working
- [x] Ran and adapted existing dhcp non nat tests

## Tasks

- [x] Get integ test fully working
- [x] Implement renew
- [x] Add renew test

## Test Plan

Commit:
chore(agw): Dhcp cli integration

* Started to use the dhcp cli in the ip_allocator to seperate dhcp and mobilityd functionallity.
* Add integ test for dhcp and non nat mode.

Signed-off-by: Christian Krämer <christian.kraemer@tngtech.com>

Commit:
chore(agw): Add release ip block

Signed-off-by: Christian Krämer <christian.kraemer@tngtech.com>

Commit:
fix(agw): fix dhcp cli calls

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
refactor(agw): Added tests for ip allocator

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
chroe(ci): Convert tests from unittest to pytest style

Signed-off-by: Christian Krämer <christian.kraemer@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Removed timings from tests

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

Commit:
chore(agw): add unit tests for ip allocator dhcp

Add tests for removing ip blocks

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Refactored tests

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

Commit:
chore(agw): add unit tests for alloc_ip_address

* some linting fixes
* the last assert still does not work, there are no entries of allocated
  ips in the ip_state_map

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

mpfirrmann commented on lte/gateway/deploy/roles/trfserver/tasks/main.yml:
```diff
@@ -98,6 +97,31 @@
     - iperf3_3.9-1_amd64.deb
   when: preburn
 
+- name: configure dhcpd
```
Could these steps be moved to preburn? If so, please add a `TODO` comment.

crasu replied:
done

mpfirrmann commented on lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_non_nat_dhcp.py:
```diff
@@ -0,0 +1,66 @@
+"""
+Copyright 2020 The Magma Authors.
```
```suggestion
Copyright 2022 The Magma Authors.
```

crasu replied:
done

mpfirrmann commented on lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_non_nat_dhcp.py:
```diff
@@ -0,0 +1,66 @@
+from s1ap_utils import MagmadUtil
+
+
+class TestAttachDetach(unittest.TestCase):
```
```suggestion
class TestAttachDetachWIthNonNatDhcp(unittest.TestCase):
```

crasu replied:
done

mpfirrmann commented on lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_non_nat_dhcp.py:
```diff
@@ -0,0 +1,66 @@
+        self.magma_utils.disable_dhcp_config()
+        self.magma_utils.enable_nat()
+
+    def test_attach_detach(self):
```
See above.
```suggestion
    def test_attach_detach_with_non_nat_dhcp(self):
```

mpfirrmann commented on lte/gateway/python/magma/mobilityd/tests/test_dhcp_helper_cli.py:
```diff
@@ -0,0 +1,131 @@
+    assert pkt[Ether].src == MACSTRING.lower()
+    assert Dot1Q not in pkt
+    assert ('message-type', 'discover') in pkt[DHCP].options
+
```
Do we want to assert the DHCP state here as well?
```suggestion
assert dhcp_helper_cli_fixture._state == DHCPState.DISCOVER

```

crasu replied:
done

mpfirrmann commented on lte/gateway/python/scripts/dhcp_helper_cli.py:
```diff
@@ -0,0 +1,353 @@
+
+    def send_dhcp_pkt(self, dhcp_opts: List[Any], pkt_xid: int, ciaddr: Optional[str] = None) -> None:
+        pkt = Ether(src=str(self._mac), dst="ff:ff:ff:ff:ff:ff")
+        if self._vlan and self._vlan != 0:
```
`self._vlan` is int, so this may be a double check for != 0.

crasu replied:
How much do we trust those type annotations?

Commit:
chore(agw): Update tests use fixtures instead of functions unify them whereever possible

Signed-off-by: Christian Krämer <christian.kraemer@tngtech.com>

Commit:
chore(agw): linter fixes

Signed-off-by: Christian Krämer <christian.kraemer@tngtech.com>

Commit:
refactor(agw): Refactored and fixed tests

- Renamed `DHCP_CLI_HELPER_PATH` to `DHCP_HELPER_CLI_PATH`
- Refactored som tests to use set literals
- Fixed a typo in udhcpd.conf
- Removed some debug print statements
- Refactored and extracted some test code to reduce redundancy
- Added line to `ip_allocato_dhcp.py` so that we are actually setting the new deadlines in the dhcp descriptor.

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
refactor(agw): refactor ip_allocator_dhcp

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
chore(agw): precommit linter

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

mpfirrmann commented on lte/gateway/deploy/roles/trfserver/files/udhcpd.conf:
```diff
@@ -0,0 +1,11 @@
+start           192.168.128.20  #default: 192.168.0.20
```
Do we need a license header here?

wolfseb replied:
Added.

mpfirrmann commented on lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_non_nat_dhcp.py:
```diff
@@ -0,0 +1,66 @@
+from s1ap_utils import MagmadUtil
+
+
+class TestAttachDetachWIthNonNatDhcp(unittest.TestCase):
```
Capitalization typo.
```suggestion
class TestAttachDetachWithNonNatDhcp(unittest.TestCase):
```

wolfseb replied:
Done.

mpfirrmann commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -37,14 +42,19 @@
         self, store: MobilityStore, retry_limit: int = 300,
-        iface: str = "dhcp0",
+        iface: str = "eth2", lease_renew_wait_min: float = LEASE_RENEW_WAIT_MIN,
+        start: bool = True,  # TODO read this from config file
```
Is this TODO for a follow up story?

wolfseb replied:
Probably, yes. It is pretty low priority for now to extract this into the config file.

sebathomas replied:
Why make this configurable? Does it make sense to run the IPAllocator without the monitor thread?

wolfseb replied:
We need this for testing, where we need control over when the monitor thread runs to avoid timing issues with it

sebathomas replied:
But if it's only useful for testing, we shouldn't make it configurable. I would even prefer something like `disable_monitor_thread_for_testing = False`.

sebathomas replied:
Ah, looking at the history the TODO comment is actually for the other parameters.

sebathomas replied:
Please move the comment to the parameter that should be configurable (or delete it).

mpfirrmann commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -61,14 +71,80 @@ def __init__(
+        while True:
+            wait_time = self._lease_renew_wait_min
+            with self.dhcp_wait:
+                # dhcp_desc: DHCPDescriptor
```
Remove comment?

wolfseb replied:
Done.

mpfirrmann commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -77,15 +153,92 @@ def add_ip_block(self, ipblock: IPNetwork):
-        ipblocks: List[IPNetwork],
-        force: bool = False,
+            self, ipblocks: List[IPNetwork],
+            force: bool = False,
```
The indentation does not match here.

wolfseb replied:
Done.

mpfirrmann commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -77,15 +153,92 @@ def add_ip_block(self, ipblock: IPNetwork):
+        Returns a set of the blocks that have been successfully removed.
+        """
+
+        remove_blocks = set(ipblocks) & self._store.assigned_ip_blocks
```
Out of curiosity: How does the bitwise operations work here?

wolfseb replied:
It is a set operation, not a bitwise one: it returns the intersection of the two sets

mpfirrmann commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -77,15 +153,92 @@ def add_ip_block(self, ipblock: IPNetwork):
+                "Cannot remove unknown IP block(s): %s",
+                extraneous_blocks,
+            )
+        del extraneous_blocks
```
Should this be inside the `if` statement?

wolfseb replied:
It probably doesn't matter, it just removes the variable otherwise (which should just be empty anyway if it doesn't go into the `if` clause)

mpfirrmann commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -123,32 +295,86 @@ def alloc_ip_address(self, sid: str, vlan_id: int) -> IPDesc:
+            dhcp_json = json.loads(dhcp_response.stdout)
+        except JSONDecodeError as e:
+            logging.error(
+                f"Could not decode '{dhcp_response.stdout}' received '{dhcp_response.stderr}' from dhcp_helper_cli called with parameters '{dhcp_response.args}'",
```
Why are there single quotes around the variable in the f-string?

mpfirrmann commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -123,32 +295,86 @@ def alloc_ip_address(self, sid: str, vlan_id: int) -> IPDesc:
+                subnet=str(IPv4Network(dhcp_json["subnet"], strict=False)),
+                server_ip=ip_address(dhcp_json["server_ip"]) if dhcp_json["server_ip"] else None,
+                router_ip=ip_address(dhcp_json["server_ip"]) if dhcp_json["server_ip"] else None,
+                # TODO extract router ip from dhcp pkt
```
Is this TODO for a follow up story?

wolfseb replied:
This is now implemented.

mpfirrmann commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -123,32 +295,86 @@ def alloc_ip_address(self, sid: str, vlan_id: int) -> IPDesc:
+        )
+        if dhcp_response.returncode != 0:
+            logging.error(
+                f"Could not decode '{dhcp_response.stdout}' received '{dhcp_response.stderr}' from {DHCP_HELPER_CLI_PATH} called with parameters '{dhcp_response.args}'",
```
Single quotes, see above.

mpfirrmann commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -186,35 +410,37 @@ def release_ip(self, ip_desc: IPDesc):
-                    (dhcp_desc is None) or not dhcp_allocated_ip(dhcp_desc)
+            if dhcp_cli_response.returncode != 0:
+                logging.error(
+                    f"Could not decode '{dhcp_cli_response.stdout}' received '{dhcp_cli_response.stderr}' from {DHCP_HELPER_CLI_PATH} called with parameters '{dhcp_cli_response.args}'",
```
Single quotes, see above.

mpfirrmann commented on lte/gateway/python/magma/mobilityd/tests/test_ip_allocator_dhcp.py:
```diff
@@ -0,0 +1,316 @@
+@pytest.fixture
+def ip_allocator_dhcp_fixture(
+        ip_allocator_fixture: IPAllocatorDHCP,
+        dhcp_desc_fixture: DHCPDescriptor,
```
Indentation.

wolfseb replied:
Done.

mpfirrmann commented on lte/gateway/python/magma/mobilityd/tests/test_ip_allocator_dhcp.py:
```diff
@@ -0,0 +1,316 @@
+#     m.returncode = 0
+#     m.stdout = """{"lease_expiration_time": %s}""" % LEASE_EXPIRATION_TIME
+#     return m
+
```
Is this block relevant or can it be deleted?

wolfseb replied:
It was deleted, thanks for spotting it!

mpfirrmann commented on lte/gateway/python/magma/mobilityd/tests/test_ip_allocator_dhcp.py:
```diff
@@ -0,0 +1,316 @@
+def run_dhcp_allocator_thread(
+        frozen_datetime: Any,
+        ip_allocator_dhcp_fixture: IPAllocatorDHCP,
+        freeze_time: float,
```
Indentation.

wolfseb replied:
Done.

mpfirrmann commented on lte/gateway/python/magma/mobilityd/tests/test_ip_allocator_dhcp.py:
```diff
@@ -0,0 +1,316 @@
+def test_allocate_ip_address(
+        ip_allocator_fixture: IPAllocatorDHCP,
+        ip_desc_fixture: IPDesc,
+        dhcp_desc_fixture: DHCPDescriptor,
```
Indentation.

wolfseb replied:
Done.

mpfirrmann commented on lte/gateway/python/magma/mobilityd/tests/test_ip_allocator_dhcp.py:
```diff
@@ -0,0 +1,316 @@
+
+def test_renewal_of_ip(
+        ip_allocator_dhcp_fixture: IPAllocatorDHCP,
+) -> None:
```
Indentation

wolfseb replied:
Done.

mpfirrmann commented on lte/gateway/python/magma/mobilityd/tests/test_ip_allocator_dhcp.py:
```diff
@@ -0,0 +1,316 @@
+
+@patch("subprocess.run", MagicMock())
+def test_keep_ip_block_with_allocated_ip(
+        ip_allocator_block_fixture: IPAllocatorDHCP,
```
Indentation.

wolfseb replied:
Done.

mpfirrmann commented on lte/gateway/python/magma/mobilityd/tests/test_ip_allocator_dhcp.py:
```diff
@@ -0,0 +1,316 @@
+@patch("subprocess.run", MagicMock())
+def test_force_remove_ip_block_with_allocated_ip(
+        ip_allocator_block_fixture: IPAllocatorDHCP,
+        ip_desc_fixture: IPDesc,
```
Indentation.

wolfseb replied:
Done.

Commit:
chore(agw): remove bazel build for removed tests

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
chore(agw): fix formats and linter

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
chore(agw): implement option to receive router_ip

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
chore(agw): add license header to udhcpd.conf

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Made existing tests work

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Properly stop monitor thread

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

Commit:
chore(agw): rename test file to clarify what it is testing

Also makes the monitoring thread a daemon thread, so we terminate it
when the program ends (for whichever reason)

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
WIP: add e2e tests for ip allocator dhcp with vlan

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

sebathomas commented on lte/gateway/deploy/roles/trfserver/tasks/main.yml:
```diff
@@ -98,6 +97,32 @@
     - iperf3_3.9-1_amd64.deb
   when: preburn
 
+# ODO Can be moved to preburn
```
```suggestion
# TODO Can be moved to preburn
```

wolfseb replied:
Done.

sebathomas commented on lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py:
```diff
@@ -1758,7 +1788,7 @@ def remove_ip_blocks(self, blocks):
                 objects representing the removed IP blocks.
         """
-        removed_blocks = self._mobility_client.remove_ip_blocks(blocks)
+        removed_blocks = self._mobility_client.remove_ip_blocks(blocks, force=True)
```
Why do we need the `force`?

wolfseb replied:
This is for test cleanup (this function is only used in  `MobilityUtil.cleanup`), so that all allocated IPs and IP blocks are removed at the end of each test no matter what (otherwise some stuff will be left in the redis)

sebathomas commented on lte/gateway/python/magma/mobilityd/tests/test_ip_allocator_dhcp_with_vlan.py:
```diff
@@ -1,9 +1,7 @@
-
 This source code is licensed under the BSD-style license found in the
 LICENSE file in the root directory of this source tree.
-
```
?

wolfseb replied:
!

MoritzThomasHuebner committed:
refactor(agw): Fixed tests

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Minor refactors, removed variable xid

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

Commit:
chore(agw): cleanup

* Fixing issue with the monitor_thread in ip_allocator_dhcp tests
* Refactoring in ip_allocator_dhcp
* Precommit linter

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

nstng commented on lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_non_nat_dhcp.py:
```diff
@@ -0,0 +1,66 @@
+"""
```
Does this test need to be added to `lte/gateway/python/integ_tests/defs.mk`?

wolfseb replied:
Probably yes. I guess `precommit` would make the most sense to put it into?

nstng replied:
if it's sufficiently fast I would add it as a precommit test.

nstng commented on lte/gateway/python/magma/mobilityd/tests/BUILD.bazel:
```diff
@@ -20,9 +20,9 @@ ORC8R_ROOT = "{}orc8r/gateway/python".format(MAGMA_ROOT)
 LTE_ROOT = "{}lte/gateway/python".format(MAGMA_ROOT)
```
Do we also need to add/rename tests in `lte/gateway/python/defs.mk`?

wolfseb replied:
Good point!

nstng commented on lte/gateway/python/magma/mobilityd/tests/test_ip_allocator_dhcp.py:
```diff
@@ -0,0 +1,310 @@
+"""
```
In bazel this is marked as a sudo test - should the test cases be annotated with something like `@unittest.skipIf(os.getuid(), reason="needs root user")`?

wolfseb replied:
I think we don't actually need sudo for these tests, as the script that would require it is mocked. I removed the flag in bazel.

nstng commented on lte/gateway/python/scripts/dhcp_helper_cli.py:
```diff
@@ -75,16 +76,20 @@ class DhcpHelperCli:
     _TIMEOUT = 10
     _state = DHCPState.DISCOVER
-    _msg_xid = 0
+    _msg_xid = random.randint(0, 2**32 - 1)
```
Why is random needed here? Why 2^32-1?

wolfseb replied:
On every call of the `dhcp_helper_cli`, a new instance gets created. If we leave the `_msg_xid` on any given number here, it will always be the same, and we cannot discern between different packets with it (i.e. the `_msg_xid` will be rather pointless in that case). We take a random integer to solve this issue. We chose 2^32-1 because the xid in the scapy `BOOTP` packet is a 32 bit field (note that the `random.randint` function is inclusive of the upper bound).

nstng commented on lte/gateway/python/scripts/dhcp_helper_cli.py:
```diff
@@ -107,7 +112,7 @@ def allocate(self):
     def release(self):
         self.send_dhcp_release()
-        self.wait_for(self.receive_dhcp_release)
+        # Receiving release ack is not mandatory
```
Does this comment describe why `self.wait_for(self.receive_dhcp_release)` was removed? I.e., does this comment only make sense if someone is viewing the diff? Then I would remove it.

wolfseb replied:
I would say it describes why we don't wait for receiving an ack response as compared to the methods `allocate` and `renew`

Commit:
chore(agw): add new tests to defs.mk and BUILD.bazel

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

sebathomas commented on lte/gateway/python/integ_tests/defs.mk:
```diff
@@ -118,7 +118,8 @@ s1aptests/test_attach_nw_initiated_detach_with_mme_restart.py \
 s1aptests/test_attach_ul_udp_data_with_sessiond_restart.py \
-s1aptests/test_attach_detach_setsessionrules_tcp_data.py
+s1aptests/test_attach_detach_setsessionrules_tcp_data.py \
+s1aptests/test_attach_detach_with_non_nat_dhcp.py \
```
Nit: the trailing backslash is inconsistent with the other groups.

wolfseb replied:
> * Is there another way to test the feature on my machine except running the integration test? Is it normal that this test is very slow? (It takes over 10 minutes and it seems that not just restarts but also operations like a detach take a while.)

For me it only takes around 4.5 minutes. But the restart of mobilityd at the end is painfully slow (and the timeout does not work either, it just waits until mobilityd is up anyway). There is also the issue with `magma_dp@envoy` not restarting, which also happens on master (here it waits a long time as well, but the `envoy` does not come up at all. I don't know where that wait time comes from). The rest (the actual test) just takes a few seconds.

sebathomas replied:
Today I got a green run in 2 minutes after rebuilding my VM. I think my setup yesterday was not behaving normally.

wolfseb replied:
We also fixed the threading which caused mobilityd to take very long to restart ;)

sebathomas commented on lte/gateway/python/magma/mobilityd/tests/test_ip_allocator_dhcp_with_vlan.py:
```diff
@@ -91,71 +89,71 @@ def test_dhcp_lease1(self):
+        self._validate_ip_subnet(sid1, vlan1)
+        self._release_ip(sid1, vlan1, dhcp_desc)
 
     @unittest.skip("needs more investigation.")
```
Why did you edit the test if it's still skipped? Can it be run now?

wolfseb replied:
We just adapted the logic to the changed framework, similar to the other tests. It still does not work.

MoritzThomasHuebner replied:
Should we just delete it then? From what we can tell it never worked.

sebathomas replied:
Let's leave it.

sebathomas commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -61,14 +71,90 @@ def __init__(
+            self._monitor_thread = threading.Thread(
+                target=self._monitor_dhcp_state,
+            )
+            self._monitor_thread.daemon = True
```
Can also be set directly in the Thread constructor.

MoritzThomasHuebner replied:
We fixed an issue by also setting this in the thread constructor. For all of our tests to work, we need to be able to do a `stop_monitor_thread` and a `start_monitor_thread`. `Thread.start` can only be called once, so we need to create a new instance of Thread every time and set `daemon = True`

sebathomas replied:
What I meant is you can set the flag when you create the thread.
```
self._monitor_thread = threading.Thread(
    target=self._monitor_dhcp_state, daemon=True
)
```
(It's not important though.)

sebathomas reviewed (commented):
I think the PR description could be improved.
- Can you write a bit more in the summary?
- Is there another way to test the feature on my machine except running the integration test? Is it normal that this test is very slow? (It takes over 10 minutes and it seems that not just restarts but also operations like a detach take a while.)

MoritzThomasHuebner committed:
fix(agw): Sped up the integ test

- Fixed an issue where we did not set the monitor thread to be in daemon mode
- Adapted test code
- Removed a superfluous backslash

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

sebathomas commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -61,14 +71,93 @@ def __init__(
+        if join:
+            self._monitor_thread.join()
+        if reset:
+            self._monitor_thread = None
```
You can fix this if you give it a type `Optional[threading.Thread]` in line 77. Then you'll need another null check in `stop_monitor_thread`.

sebathomas commented on lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_non_nat_dhcp.py:
```diff
@@ -0,0 +1,66 @@
+
+    def tearDown(self):
+        """Cleanup after test case execution"""
+        self._s1ap_wrapper.cleanup()
```
I just had  the case where the test failed and didn't clean up the dhcp and nat config, not sure why it happened. Is it possible this cleanup method fails and then the steps after that are not run?

sebathomas replied:
But even if it does not fail, it might be better to adapt the mconfig first and then do the rest of the cleanup and restart all services.

sebathomas committed:
test: Use two UEs for DHCP integration test

Signed-off-by: Sebastian Thomas <sebastian.thomas@tngtech.com>

MoritzThomasHuebner committed:
fix(agw): Minor fix-ups discovered during review

- Reduced xid logic
- Re-arranged logic in integ-test cleanup
- Minor typing issue in ip_allocator_dhcp.py
- Some weird import

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

sebathomas committed:
test: adapt comments

Signed-off-by: Sebastian Thomas <sebastian.thomas@tngtech.com>

sebathomas committed:
test: remove xid from mock

Signed-off-by: Sebastian Thomas <sebastian.thomas@tngtech.com>

Commit:
fix(agw): move todo to clarify its context

* move the todo in ip_allocator_dhcp
* move daemon=true into the thread constructor for the monitor thread in
  IPAllocatorDHCP

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

sebathomas commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -37,14 +42,18 @@
 DEFAULT_DHCP_REQUEST_RETRY_DELAY = 1
+LEASE_RENEW_WAIT_MIN = 200
 
+DHCP_HELPER_CLI_PATH = f"{environ.get('MAGMA_ROOT')}/lte/gateway/python/scripts/dhcp_helper_cli.py"
```
Oh, I missed this. This is fine for dev but I don't think that's going to work with deployed installations of magma. They currently install the script at `/usr/local/bin/dhcp_helper_cli.py` and don't rely on `MAGMA_ROOT`.

I am not sure what's the best way to fix it, but we should focus on deployments first and only allow dev as a backup. Maybe check if `dhcp_helper_cli.py` is in the user's `PATH` and only try the `MAGMA_ROOT` path as a backup.

sebathomas replied:
Maybe we don't even need the fallback. I think it's in the PATH in magma-dev too.

wolfseb replied:
It is not in the `PATH` for me, actually. But that should be easy to fix in the meantime, but we can just need a link in the meantime

sebathomas replied:
I tested that it works for and will update the file to call the script directly. (I think you need to be in the virtualenv to see it in the PATH).

Commit:
chore(agw): add dhcp_helper_cli to bazel lte scripts

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

Commit:
fix(agw): fix dhcp integ test

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Fixed a typo

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

sebathomas committed:
fix(agw): Don't use magma repo path to call CLI tool

Signed-off-by: Sebastian Thomas <sebastian.thomas@tngtech.com>

sebathomas committed:
fix(agw): Formatting

Signed-off-by: Sebastian Thomas <sebastian.thomas@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Reverted helper cli path

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Added to-do

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
refactor(agw): Added logic to select scripts

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

Commit:
fix(agw): make pylint happy

Signed-off-by: Sebastian Wolf <sebastian.wolf@tngtech.com>

sebathomas commented on lte/gateway/python/magma/mobilityd/ip_allocator_dhcp.py:
```diff
@@ -37,14 +43,22 @@
+LEASE_RENEW_WAIT_MIN = 200
 
+# TODO: Change this back to just "dhcp_helper_cli.py" in PR 14635
+if os.path.exists(f"{environ.get('MAGMA_ROOT')}/lte/gateway/python/scripts/dhcp_helper_cli.py"):
```
Nitpick: you might want to check the existence of the env variable first like `if environment.get('MAGMA_ROOT') and os.path.exists(...)`. Otherwise this will look for the path "None/lte/gateway/python/scripts/dhcp_helper_cli.py".

sebathomas commented on lte/gateway/deploy/roles/trfserver/files/udhcpd.conf:
```diff
@@ -0,0 +1,23 @@
+start           192.168.128.20  #default: 192.168.0.20
+end             192.168.128.254 #default: 192.168.0.254
+interface       eth2            #default: eth0
+opt     dns     192.168.10.2 192.168.10.10
```
I think these IPs were copied from an example config and don't exist in the typical magma-dev setup. This results in ARP traffic containing these IPs when udhcpd is being used, which can confuse pipelined. However, these errors in pipelined don't appear if you revert #14405, so this should be fixed in pipelined.

It would be good to get an easier way to reproduce it, and and to track it in a bug ticket.

```
Dec 13 12:59:41 magma-dev pipelined[132217]: WARNING:root:Unexpected IP in ARP response. expected: 192.168.10.2 pkt: Ethernet header
Dec 13 12:59:41 magma-dev pipelined[132217]:   dst_mac : ff:ff:ff:ff:ff:ff
Dec 13 12:59:41 magma-dev pipelined[132217]:   src_mac : 08:00:27:2b:fc:6d
Dec 13 12:59:41 magma-dev pipelined[132217]:   ptype   : IPv4
Dec 13 12:59:41 magma-dev pipelined[132217]: ARP packet
Dec 13 12:59:41 magma-dev pipelined[132217]:   hwtype : 0x1194
Dec 13 12:59:41 magma-dev pipelined[132217]:   ptype  : unknown
Dec 13 12:59:41 magma-dev pipelined[132217]:   hwlen  : 0
Dec 13 12:59:41 magma-dev pipelined[132217]:   plen   : 0
Dec 13 12:59:41 magma-dev pipelined[132217]:   op     : unknown
Dec 13 12:59:41 magma-dev pipelined[132217]:   hwsrc  : 40:11:37:d3:c0:a8
Dec 13 12:59:41 magma-dev pipelined[132217]:   psrc   : 129.42.255.255
Dec 13 12:59:41 magma-dev pipelined[132217]:   hwdst  : ff:ff:00:43:00:44
Dec 13 12:59:41 magma-dev pipelined[132217]:   pdst   : 1.52.140.206
```

sebathomas replied:
Debug logging showed that pipelined is actually sending a packet with that gw ip:
```
Dec 13 16:11:28 magma-dev pipelined[219512]: DEBUG:root:_get_gw_mac_address_v4 192.168.10.2 NO_VLAN uplink_br0
Dec 13 16:11:28 magma-dev pipelined[219512]: DEBUG:root:sending arp via egress: uplink_br0
Dec 13 16:11:28 magma-dev pipelined[219512]: DEBUG:root:ARP Req pkt:
Dec 13 16:11:28 magma-dev pipelined[219512]: None
Dec 13 16:11:32 magma-dev pipelined[219512]: DEBUG:root:ARP Res pkt Ethernet header
Dec 13 16:11:32 magma-dev pipelined[219512]:   dst_mac : 08:00:27:2b:fc:6d
Dec 13 16:11:32 magma-dev pipelined[219512]:   src_mac : 08:00:27:7d:89:93
Dec 13 16:11:32 magma-dev pipelined[219512]:   ptype   : IPv4
Dec 13 16:11:32 magma-dev pipelined[219512]: ARP packet
Dec 13 16:11:32 magma-dev pipelined[219512]:   hwtype : 0x1194
Dec 13 16:11:32 magma-dev pipelined[219512]:   ptype  : unknown
Dec 13 16:11:32 magma-dev pipelined[219512]:   hwlen  : 0
Dec 13 16:11:32 magma-dev pipelined[219512]:   plen   : 99
Dec 13 16:11:32 magma-dev pipelined[219512]:   op     : unknown
Dec 13 16:11:32 magma-dev pipelined[219512]:   hwsrc  : 40:11:e8:8d:c0:a8
Dec 13 16:11:32 magma-dev pipelined[219512]:   psrc   : 129.1.8.8
Dec 13 16:11:32 magma-dev pipelined[219512]:   hwdst  : 08:08:90:0c:00:35
Dec 13 16:11:32 magma-dev pipelined[219512]:   pdst   : 0.47.81.250
Dec 13 16:11:32 magma-dev pipelined[219512]: WARNING:root:Unexpected IP in ARP response. expected: 192.168.10.2 pkt: Ethernet header
Dec 13 16:11:32 magma-dev pipelined[219512]:   dst_mac : 08:00:27:2b:fc:6d
Dec 13 16:11:32 magma-dev pipelined[219512]:   src_mac : 08:00:27:7d:89:93
Dec 13 16:11:32 magma-dev pipelined[219512]:   ptype   : IPv4
Dec 13 16:11:32 magma-dev pipelined[219512]: ARP packet
Dec 13 16:11:32 magma-dev pipelined[219512]:   hwtype : 0x1194
Dec 13 16:11:32 magma-dev pipelined[219512]:   ptype  : unknown
Dec 13 16:11:32 magma-dev pipelined[219512]:   hwlen  : 0
Dec 13 16:11:32 magma-dev pipelined[219512]:   plen   : 99
Dec 13 16:11:32 magma-dev pipelined[219512]:   op     : unknown
Dec 13 16:11:32 magma-dev pipelined[219512]:   hwsrc  : 40:11:e8:8d:c0:a8
Dec 13 16:11:32 magma-dev pipelined[219512]:   psrc   : 129.1.8.8
Dec 13 16:11:32 magma-dev pipelined[219512]:   hwdst  : 08:08:90:0c:00:35
Dec 13 16:11:32 magma-dev pipelined[219512]:   pdst   : 0.47.81.250
```

In the new implementation, it logs an error and complains about the unexpected IP, in the old implementation there is no response.

sebathomas replied:
(We might just remove the router and dns lines from the config though, doesn't seem to break anything in the test and avoids the pipelined error. But the error seems to indicate a real issue.)

MoritzThomasHuebner committed:
refactor(agw): improved a path check

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>
